File: ext/src/core/draw.cpp
// Copyright (c) 2016, Herman Bergwerf. All rights reserved.
// Use of this source code is governed by a MIT-style license
// that can be found in the LICENSE file.

#include "draw.hpp"

namespace fldart {
FunctionMapping _draw::methods[] = {
  {"fldart::color", color},
  {"fldart::line1", line1},
  {"fldart::line2", line2},
  {NULL, NULL}
};

void color(Dart_NativeArguments arguments) {
  int64_t c;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &c));
  fl_color(static_cast<Fl_Color>(c));
  Dart_Handle result = Dart_Null();
  Dart_SetReturnValue(arguments, result);
  Dart_ExitScope();
}

void line1(Dart_NativeArguments arguments) {
  int64_t x;
  int64_t y;
  int64_t x1;
  int64_t y1;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &x));
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 1)), &y));
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 2)), &x1));
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 3)), &y1));
  fl_line(x,y,x1,y1);
  Dart_Handle result = Dart_Null();
  Dart_SetReturnValue(arguments, result);
  Dart_ExitScope();
}

void line2(Dart_NativeArguments arguments) {
  int64_t x;
  int64_t y;
  int64_t x1;
  int64_t y1;
  int64_t x2;
  int64_t y2;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &x));
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 1)), &y));
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 2)), &x1));
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 3)), &y1));
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 4)), &x2));
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 5)), &y2));
  fl_line(x,y,x1,y1,x2,y2);
  Dart_Handle result = Dart_Null();
  Dart_SetReturnValue(arguments, result);
  Dart_ExitScope();
}
}

File: lib/src/fl/draw.dart
// Copyright (c) 2016, Herman Bergwerf. All rights reserved.
// Use of this source code is governed by a MIT-style license
// that can be found in the LICENSE file.

part of fltk;

/// Bindings for fl_draw.H

/// Set the drawing color.
set color(int c) native 'fldart::color';

/// Draw line from ([x], [y]) to ([x1], [y1]).
void _line1(int x, int y, int x1, int y1) native 'fldart::line1';

/// Draw line from ([x], [y]) to ([x1], [y1]) and then to ([x2], [y2]).
void _line2(int x, int y, int x1, int y1, int x2, int y2)
    native 'fldart::line2';

/// Draw line from ([x], [y]) to ([x1], [y1]) and optionally to ([x2], [y2]).
void line(int x, int y, int x1, int y1, [int x2 = null, int y2 = null]) {
  if (x2 != null && y2 != null) {
    _line2(x, y, x1, y1, x2, y2);
  } else {
    _line1(x, y, x1, y1);
  }
}

/// Draw image data using `fl_draw_image`.
void drawImage(Uint8List buffer, int x, int y, int w, int h,
    [int d = 3, int l = 0]) native 'fldart::draw_image';

/// Draw image data using Fl_RGB_Image and fallback alpha blending driver.
void drawRgbImage(Uint8List buffer, int x, int y, int w, int h,
    [int d = 3, int l = 0]) native 'fldart::rgb_image_draw';
